File: run_scraping_flow.js
import { spawn } from 'child_process';
import path from 'path';

// Define the steps in order
const STEPS = [
    { script: 'session_scraper.js', desc: 'Scraping Session List' },
    { script: 'playback_scraper.js', desc: 'Scraping Video Links' },
    { script: 'download_videos.js', desc: 'Downloading Raw Videos' },
    { script: 'merge_videos.js',    desc: 'Merging into MP4' }
];

async function runStep(step) {
    return new Promise((resolve, reject) => {
        console.log(`\n🔹 [Step] ${step.desc}...`);
        
        const child = spawn('node', [step.script], { 
            stdio: 'inherit',
            shell: true 
        });
        
        child.on('close', (code) => {
            if (code === 0) {
                console.log(`✅ [Step] ${step.desc} Completed`);
                resolve();
            } else {
                console.error(`❌ [Step] ${step.desc} Failed (Exit Code: ${code})`);
                reject(new Error(`Script ${step.script} failed`));
            }
        });
        
        child.on('error', (err) => {
            console.error(`❌ [Step] Error launching ${step.script}: ${err.message}`);
            reject(err);
        });
    });
}

async function main() {
    console.log("🚀 Starting Full Scraping Pipeline");
    console.log("=========================================");
    
    // Check for .env file
    // (Optional simple check, but good practice)
    
    const startTime = Date.now();
    
    try {
        for (const step of STEPS) {
            await runStep(step);
        }
        
        const duration = ((Date.now() - startTime) / 1000).toFixed(1);
        console.log("\n=========================================");
        console.log(`🎉 Pipeline Completed Successfully in ${duration}s!`);
        console.log("=========================================");
        
    } catch (error) {
        console.error("\n⛔ Pipeline Stopped due to error.");
        process.exit(1);
    }
}

main();

File: run_scraping_flow.py
import subprocess
import time
import sys
import os

# Define the steps in order
STEPS = [
    {'script': 'python_code/session_scraper.py', 'desc': 'Scraping Session List'},
    {'script': 'python_code/playback_scraper.py', 'desc': 'Scraping Video Links'},
    {'script': 'python_code/download_videos.py', 'desc': 'Downloading Raw Videos'},
    {'script': 'python_code/merge_videos.py',    'desc': 'Merging into MP4'}
]

def run_step(step):
    """Run a single step"""
    print(f"\n🔹 [Step] {step['desc']}...")
    
    try:
        # Use python3 explicit command or just python depending on env
        # We assume 'python' is the correct interpreter here
        cmd = ["python", step['script']]
        
        # Check if script exists relative to current dir
        if not os.path.exists(step['script']):
            # Try without python_code/ prefix if we are inside it
            if os.path.exists(os.path.basename(step['script'])):
                cmd = ["python", os.path.basename(step['script'])]
            else:
                print(f"❌ Script not found: {step['script']}")
                return False
        
        result = subprocess.run(cmd, check=False)
        
        if result.returncode == 0:
            print(f"✅ [Step] {step['desc']} Completed")
            return True
        else:
            print(f"❌ [Step] {step['desc']} Failed (Exit Code: {result.returncode})")
            return False
            
    except Exception as e:
        print(f"❌ [Step] Error launching {step['script']}: {e}")
        return False

def main():
    print("🚀 Starting Full Scraping Pipeline (Python)")
    print("=========================================")
    
    start_time = time.time()
    
    try:
        for step in STEPS:
            if not run_step(step):
                print("\n⛔ Pipeline Stopped due to error.")
                sys.exit(1)
        
        duration = round(time.time() - start_time, 1)
        print("\n=========================================")
        print(f"🎉 Pipeline Completed Successfully in {duration}s!")
        print("=========================================")
        
    except KeyboardInterrupt:
        print("\n\n⛔ Pipeline stopped by user.")
        sys.exit(1)

if __name__ == "__main__":
    main()
